### ISSUE TITLE
Inconsistent skolemization

### ISSUE THREAD

**Issue by jstolarek** [jstolarek/inferno#11]
At the moment when we typecheck a let binding with a signature we skolemize the quantifiers in the signature, unify it with an inferred type of the bound term, and then unskolemize the quantifiers. However, when we typecheck an annotated lambda we skolemize the quantifiers, typecheck the body, and keep the skolems. Without keeping the skolems some examples rightly fail - see #3. This is inconsistent and unsatisfying. I think the right thing to do from a theoretical standpoint would be to skolemize the quantifiers when checking the type of the annotated binder against the type of the argument in a function application, and after the check succeeds unskolemize the quantifiers. However, inferno does not allow doing something like this easily. For let bindings it generates a `CLet` constraint so it's easy to place the skolemization logic in the right place, but for function application it generates:

![app](https://user-images.githubusercontent.com/1597921/98793839-546d9600-2400-11eb-9c9e-11078ca75cff.png)

so basically an existential with a nested conjunction of constraints. With such constraints it isn't immediately obvious that we are typechecking a lambda application. However, for lambdas we generate `CDef` constraints, so *perhaps* we could peek into the structure of constraints when handling `CExist` and if the structure is `CConj (CDef ...)` we could unskolemize quantifiers on the signature. This sounds like a horrible hack but I believe it will work. The first thing I need however is an example that actually breaks with the current behaviour.

**Comment by jstolarek** [comment]
After some investigation it turns out that the right thing to do is to skolemize the quantifiers in a frozen type and unskolemize them after unification (see https://github.com/jstolarek/inferno/issues/15#issuecomment-735768387), and at the same time not to skolemize the quantifiers in an annotated lambda binder. This handles all the test cases correctly.
